data/abi/abi_type.go

```diff
@@ -0,0 +1,499 @@
+	Ufixed
+	// Bool is the index (3) for `Bool` type in ABI encoding.
+	Bool
+	// ArrayStatic is the index (4) for static length array (<type>[length]) type in ABI encoding.
+	ArrayStatic
+	// Address is the index (5) for `Address` type in ABI encoding (an type alias of Byte[32]).
+	Address
+	// ArrayDynamic is the index (6) for dynamic length array (<type>[]) type in ABI encoding.
+	ArrayDynamic
+	// String is the index (7) for `String` type in ABI encoding (an type alias of Byte[]).
+	String
+	// Tuple is the index (8) for tuple `(<type 0>, ..., <type k>)` in ABI encoding.
+	Tuple
+)
+
+// Type is the struct that stores information about an ABI value's type.
+type Type struct {
+	abiTypeID  BaseType
+	childTypes []Type
+
+	// only can be applied to `uint` bitSize <N> or `ufixed` bitSize <N>
+	bitSize uint16
+	// only can be applied to `ufixed` precision <M>
+	precision uint16
+
+	// length for static array / tuple
+	/*
+		by ABI spec, len over binary array returns number of bytes
+		the type is uint16, which allows for only lenth in [0, 2^16 - 1]
+		representation of static length can only be constrained in uint16 type
+	*/
+	// NOTE may want to change back to uint32/uint64
+	staticLength uint16
+}
+
+// String serialize an ABI Type to a string in ABI encoding.
+func (t Type) String() string {
+	switch t.abiTypeID {
+	case Uint:
+		return "uint" + strconv.Itoa(int(t.bitSize))
+	case Byte:
+		return "byte"
+	case Ufixed:
+		return "ufixed" + strconv.Itoa(int(t.bitSize)) + "x" + strconv.Itoa(int(t.precision))
+	case Bool:
+		return "bool"
+	case ArrayStatic:
+		return t.childTypes[0].String() + "[" + strconv.Itoa(int(t.staticLength)) + "]"
+	case Address:
+		return "address"
+	case ArrayDynamic:
+		return t.childTypes[0].String() + "[]"
+	case String:
+		return "string"
+	case Tuple:
+		typeStrings := make([]string, len(t.childTypes))
+		for i := 0; i < len(t.childTypes); i++ {
+			typeStrings[i] = t.childTypes[i].String()
+		}
+		return "(" + strings.Join(typeStrings, ",") + ")"
+	default:
+		panic("Type Serialization Error, fail to infer from abiTypeID (bruh you shouldn't be here)")
+	}
+}
+
+var staticArrayRegexp *regexp.Regexp = nil
+var ufixedRegexp *regexp.Regexp = nil
+var regexpInitialized = false
+
+func initRegexps() (err error) {
+	if regexpInitialized {
+		return nil
+	}
+	// Note that we allow only decimal static array length
+	staticArrayRegexp, err = regexp.Compile(`^([a-z\d\[\](),]+)\[([1-9][\d]*)]$`)
+	if err != nil {
+		return
+	}
+	ufixedRegexp, err = regexp.Compile(`^ufixed([1-9][\d]*)x([1-9][\d]*)$`)
+	if err != nil {
+		return
+	}
+	regexpInitialized = true
+	return
+}
+
+// TypeFromString de-serialize ABI type from a string following ABI encoding.
+func TypeFromString(str string) (Type, error) {
+	// if the regexps for type str matching are not initialized
+	// attempt to initialize with `initRegexps`
+	// if error return err
+	if !regexpInitialized {
+		err := initRegexps()
+		if err != nil {
+			return Type{}, err
+		}
+	}
```

Comment:
I think it's better to call this from an init() function. https://golang.org/doc/effective_go#init

Reply:
I think so.  They're static patterns, so it would mean there is a badly written regexp.